File: packages/react-native/src/bindings.ts
// Wiring the SDK into a React Native app's own machinery
// (docs/PROTOCOL.md §3, §5).
//
// This is the file that decides whether a report is worth having. An SDK that
// is only initialised still files reports; they just say nothing — no failed
// request, no logged error, no screen name, and a crash nobody sees. Each
// binding below hands the reproduction one kind of evidence it cannot get on
// its own.
//
// Everything here patches a GLOBAL the host app shares, so every function is
// written to the same three rules:
//
//  1. IT CHAINS, NEVER REPLACES. An app already has a crash reporter and a
//     console; taking one over is a regression the app author did not ask for.
//  2. IT UNINSTALLS. Each returns a teardown that restores exactly what was
//     there — including when two of ours are installed over each other.
//  3. IT NEVER THROWS INTO THE HOST. A bug in observation must not become a bug
//     in the app being observed.
import { compact } from './protocol.ts';
import type { CrashReporter } from './crash.ts';
import type { TraceRecorder } from './recorder.ts';

/** Undo a binding. Idempotent — calling it twice is not an error. */
export type Unbind = () => void;

/** What the collectors write to. Both are optional so an app can wire crash
 *  capture without the recorder, or vice versa. */
export interface Sinks {
  recorder?: TraceRecorder;
  crash?: CrashReporter;
}

// ── Failed requests ────────────────────────────────────────────────────────

/**
 * Record FAILED requests — never successful ones, never headers, never bodies,
 * never query strings.
 *
 * `fetch` is patched rather than an interceptor added because React Native has
 * no shared HTTP client: an app may use fetch, axios-on-fetch, or its own
 * wrapper, and all of them bottom out here. `XMLHttpRequest` is patched too,
 * since axios' default RN adapter uses it directly.
 */
export function bindNetwork(sinks: Sinks): Unbind {
  const g = globalThis as { fetch?: typeof fetch; XMLHttpRequest?: unknown };
  const originalFetch = g.fetch;
  if (!originalFetch) return () => {};

  const record = (method: string, url: string, status: number, ms: number) => {
    // A success is not evidence. Recording every 200 would bury the one request
    // that mattered and spend the trace's 500-event budget on noise.
    if (status >= 400 || status === 0) {
      sinks.recorder?.request(method, url, status, ms);
      sinks.crash?.noteFailedRequest(method, url, status, ms);
    }
  };

  const patched: typeof fetch = async (input, init) => {
    const started = Date.now();
    const method = (init?.method ?? (typeof input === 'object' && 'method' in input ? input.method : 'GET')) || 'GET';
    const url = typeof input === 'string' ? input : String((input as { url?: string }).url ?? input);
    try {
      const res = await originalFetch(input, init);
      record(method, url, res.status, Date.now() - started);
      return res;
    } catch (err) {
      // status 0 is a transport failure — no response ever arrived. It is the
      // single most useful line in an offline bug report.
      record(method, url, 0, Date.now() - started);
      throw err;
    }
  };

  g.fetch = patched;
  return () => {
    // Only restore if nothing else patched on top of us; clobbering a later
    // interceptor would silently disable the app's own instrumentation.
    if (g.fetch === patched) g.fetch = originalFetch;
  };
}

// ── Logged errors ──────────────────────────────────────────────────────────

/**
 * Capture `console.error` / `console.warn` — the SHAPE of what was logged, and
 * only inside the caps.
 *
 * The arguments are stringified and truncated, which is deliberate rather than
 * lazy: an app logs objects, and an object logged by an app that is mid-failure
 * is the most likely place for a user's data to appear. A message is evidence;
 * a serialized payload is a leak.
 */
export function bindConsole(sinks: Sinks): Unbind {
  const original = { error: console.error, warn: console.warn };

  const capture = (level: 'error' | 'warn') =>
    function patched(this: unknown, ...args: unknown[]) {
      try {
        const message = args.map(shapeOf).join(' ');
        sinks.recorder?.console(level, message);
        sinks.crash?.noteConsole(level, message);
      } catch {
        // Observation must never break the app being observed.
      }
      original[level].apply(this, args as never);
    };

  console.error = capture('error');
  console.warn = capture('warn');
  return () => {
    console.error = original.error;
    console.warn = original.warn;
  };
}

/**
 * What a logged argument WAS, not what it contained.
 *
 * A string contributes itself (it is a message the developer wrote). An Error
 * contributes its name and message. Everything else contributes its shape —
 * `[Object]`, `[Array(3)]` — because the alternative is serializing an object
 * an app logged while failing, which is exactly where a user's record ends up.
 */
export function shapeOf(value: unknown): string {
  if (typeof value === 'string') return value;
  if (value === null) return 'null';
  if (value === undefined) return 'undefined';
  if (value instanceof Error) return `${value.name}: ${value.message}`;
  if (Array.isArray(value)) return `[Array(${value.length})]`;
  if (typeof value === 'object') return '[Object]';
  return String(value);
}

// ── Crashes ────────────────────────────────────────────────────────────────

/** React Native's global handler hook. Structural so this file needs no
 *  react-native import and stays testable on a laptop. */
interface ErrorUtilsLike {
  getGlobalHandler?: () => ((error: Error, isFatal?: boolean) => void) | undefined;
  setGlobalHandler?: (handler: (error: Error, isFatal?: boolean) => void) => void;
}

/**
 * File uncaught errors, CHAINING to whatever handler was already installed.
 *
 * Chaining is not politeness — an app with Crashlytics or Sentry already has a
 * global handler, and replacing it would silently stop their crash reporting
 * the day this SDK is added. Ours runs first (so the report is captured even if
 * theirs tears the process down), then theirs runs unchanged.
 *
 * The report is only PERSISTED here. A network call inside a handler on a dying
 * process is a race the handler usually loses; `flush` sends it on next launch.
 */
export function bindCrashHandler(crash: CrashReporter, errorUtils?: ErrorUtilsLike): Unbind {
  const utils = errorUtils ?? (globalThis as { ErrorUtils?: ErrorUtilsLike }).ErrorUtils;
  if (!utils?.setGlobalHandler) return () => {};
  const previous = utils.getGlobalHandler?.();

  const handler = (error: Error, isFatal?: boolean) => {
    try {
      void crash.capture('error', compact({
        name: error?.name,
        message: error?.message,
        stack: error?.stack,
      }));
    } catch {
      // Never let the reporter be the reason a crash handler throws.
    }
    previous?.(error, isFatal);
  };

  utils.setGlobalHandler(handler);
  return () => {
    if (previous) utils.setGlobalHandler?.(previous);
  };
}

/**
 * File unhandled promise rejections.
 *
 * A separate kind (`unhandledrejection`) rather than folding into `error`,
 * because the two mean different things to whoever reads the report: an
 * uncaught error killed a render, a rejected promise means something the app
 * started never finished and nobody was watching.
 */
export function bindRejectionHandler(crash: CrashReporter): Unbind {
  const g = globalThis as {
    addEventListener?: (t: string, h: (e: unknown) => void) => void;
    removeEventListener?: (t: string, h: (e: unknown) => void) => void;
  };
  if (!g.addEventListener) return () => {};

  const handler = (event: unknown) => {
    const reason = (event as { reason?: unknown })?.reason;
    const error = reason instanceof Error ? reason : undefined;
    void crash.capture('unhandledrejection', compact({
      name: error?.name ?? 'UnhandledRejection',
      message: error?.message ?? shapeOf(reason),
      stack: error?.stack,
    }));
  };
  g.addEventListener('unhandledrejection', handler);
  return () => g.removeEventListener?.('unhandledrejection', handler);
}

// ── Navigation ─────────────────────────────────────────────────────────────

/**
 * Follow the app's navigation so a recorded step can say which screen it
 * happened on — and so a crash knows the route it died on.
 *
 * Takes the route NAME, not a navigation object: React Navigation, Expo Router
 * and a hand-rolled stack all expose that differently, and an SDK that only
 * works with one router is one most apps cannot use. The host calls this from
 * `NavigationContainer.onStateChange` (or its equivalent) with a name it
 * already has.
 *
 * `cause` matters more than it looks: a route guard's redirect and a deliberate
 * tap are the same event otherwise, and a reproduction that lists an app's own
 * redirect as a step sends whoever follows it looking for a control that does
 * not exist.
 */
export function makeNavigationTracker(sinks: Sinks): {
  onRouteChange: (route: string, cause?: 'user' | 'app') => void;
  currentRoute: () => string;
} {
  let current = '';
  return {
    onRouteChange(route, cause = 'user') {
      if (!route || route === current) return;
      current = route;
      sinks.recorder?.navigate(route, cause);
    },
    // Read fresh by the crash reporter on every crash — an app navigates for
    // the whole life of its process, so a route captured once at launch would
    // put every crash on the first screen.
    currentRoute: () => current,
  };
}

// ── Everything, once ───────────────────────────────────────────────────────

export interface BindOptions extends Sinks {
  /** Absent ⇒ bound. Each can be disabled where the host has its own. */
  network?: boolean;
  console?: boolean;
  crashes?: boolean;
  errorUtils?: ErrorUtilsLike;
}

/**
 * Install every binding the sinks support, and return ONE teardown.
 *
 * Teardown order is the reverse of install order, so a handler chained on top
 * of another is always removed before the one beneath it — otherwise unbinding
 * ours would leave the host's handler pointing at a function we have discarded.
 */
export function bindAll(opts: BindOptions): Unbind {
  const undo: Unbind[] = [];
  if (opts.network !== false) undo.push(bindNetwork(opts));
  if (opts.console !== false) undo.push(bindConsole(opts));
  if (opts.crashes !== false && opts.crash) {
    undo.push(bindCrashHandler(opts.crash, opts.errorUtils));
    undo.push(bindRejectionHandler(opts.crash));
  }
  return () => {
    for (const fn of undo.reverse()) {
      try {
        fn();
      } catch {
        // A teardown that throws must not strand the others.
      }
    }
    undo.length = 0;
  };
}

File: packages/flutter/lib/src/bindings.dart
// Wiring the SDK into a Flutter app's own machinery (docs/PROTOCOL.md §3, §5).
//
// The Dart half of packages/react-native/src/bindings.ts, and deliberately the
// same three rules, because they are properties of being a guest in someone
// else's process rather than facts about a language:
//
//  1. IT CHAINS, NEVER REPLACES. An app already has a crash reporter and an
//     error handler; taking one over is a regression the app author did not ask
//     for, and it is silent.
//  2. IT UNINSTALLS. Each binding returns a teardown that restores exactly what
//     it found.
//  3. IT NEVER THROWS INTO THE HOST. A bug in observation must not become a bug
//     in the app being observed.
library;

import 'dart:async';
import 'package:flutter/foundation.dart';
import 'package:flutter/widgets.dart';

import 'crash.dart';
import 'protocol.dart';
import 'recorder.dart';

/// Undo a binding. Idempotent — calling it twice is not an error.
typedef Unbind = void Function();

/// What the collectors write to. Both optional, so an app can wire crash
/// capture without the recorder or the other way round.
class Sinks {
  const Sinks({this.recorder, this.crash});
  final TraceRecorder? recorder;
  final CrashReporter? crash;
}

// ── Logged errors ──────────────────────────────────────────────────────────

/// Capture what the app logs through `debugPrint`.
///
/// Flutter has no `console.error`, and the closest thing every app shares is
/// `debugPrint`. Only the SHAPE of a line survives the caps — a message is
/// evidence, and an object an app printed while it was failing is where a
/// user's record ends up.
Unbind bindDebugPrint(Sinks sinks) {
  final original = debugPrint;
  void patched(String? message, {int? wrapWidth}) {
    try {
      if (message != null && message.isNotEmpty) {
        sinks.recorder?.console('warn', message);
        sinks.crash?.noteConsole('warn', message);
      }
    } catch (_) {
      // Observation must never break the app being observed.
    }
    original(message, wrapWidth: wrapWidth);
  }

  debugPrint = patched;
  return () {
    if (debugPrint == patched) debugPrint = original;
  };
}

// ── Crashes ────────────────────────────────────────────────────────────────

/// File errors Flutter caught in its own framework — the `boundary` kind.
///
/// `FlutterError.onError` is where a widget build failure lands, and it is the
/// channel that matters most: the red screen (or the blank one in release) is
/// the most common fatal a Flutter app has, and NOTHING else reports it. The
/// previous handler is called afterwards, so an app with Crashlytics keeps it.
Unbind bindFlutterErrors(CrashReporter crash) {
  final previous = FlutterError.onError;
  void handler(FlutterErrorDetails details) {
    try {
      unawaited(crash.capture(
        kind: 'boundary',
        name: details.exception.runtimeType.toString(),
        message: details.exceptionAsString(),
        stack: details.stack?.toString(),
      ));
    } catch (_) {
      // Never let the reporter be the reason an error handler throws.
    }
    previous?.call(details);
  }

  FlutterError.onError = handler;
  return () {
    if (FlutterError.onError == handler) FlutterError.onError = previous;
  };
}

/// File errors that escaped the framework entirely — the `error` kind.
///
/// `PlatformDispatcher.onError` catches what `FlutterError.onError` cannot: an
/// exception thrown outside a widget lifecycle, in a callback or a stray async
/// gap. Returning the previous handler's answer (or `false`) preserves the
/// app's own decision about whether the error was handled — answering `true`
/// on its behalf would silently swallow crashes it wanted to see.
Unbind bindPlatformErrors(CrashReporter crash) {
  final previous = PlatformDispatcher.instance.onError;
  bool handler(Object error, StackTrace stack) {
    try {
      unawaited(crash.capture(
        kind: 'error',
        name: error.runtimeType.toString(),
        message: error.toString(),
        stack: stack.toString(),
      ));
    } catch (_) {
      // As above.
    }
    return previous?.call(error, stack) ?? false;
  }

  PlatformDispatcher.instance.onError = handler;
  return () {
    if (PlatformDispatcher.instance.onError == handler) {
      PlatformDispatcher.instance.onError = previous;
    }
  };
}

// ── Navigation ─────────────────────────────────────────────────────────────

/// Follows the app's navigation so a step can say which screen it happened on,
/// and a crash knows the route it died on.
///
/// A `NavigatorObserver` rather than a route-name callback, because Flutter
/// gives one interface every router honours — `MaterialApp.navigatorObservers`,
/// go_router's `observers`, auto_route's. The RN side takes a name instead, for
/// the opposite reason: React Native has no such shared interface.
///
/// `didPop` reports `cause: 'user'` and `didReplace` reports `'app'`: a pop is
/// something the reporter did, a replace is almost always a guard or a redirect
/// — and a reproduction that lists an app's own redirect as a step sends
/// whoever follows it looking for a control that does not exist.
class AlgoNavigatorObserver extends NavigatorObserver {
  AlgoNavigatorObserver(this.sinks);

  final Sinks sinks;
  String _current = '';

  /// Read FRESH by the crash reporter on every crash. An app navigates for the
  /// whole life of its process, so a route captured once at launch would put
  /// every crash on the first screen.
  String get currentRoute => _current;

  void _to(Route<dynamic>? route, String cause) {
    final name = _nameOf(route);
    if (name == null || name == _current) return;
    _current = name;
    sinks.recorder?.navigate(name, cause: cause);
  }

  /// A route's name, or null when it has none. Unnamed routes are SKIPPED
  /// rather than labelled `<unnamed>`: a reproduction step naming a screen that
  /// does not exist in the app's source is worse than one that says nothing.
  static String? _nameOf(Route<dynamic>? route) {
    final name = route?.settings.name;
    if (name == null || name.isEmpty) return null;
    return clampText(name, kTraceMaxPathLen);
  }

  @override
  void didPush(Route<dynamic> route, Route<dynamic>? previousRoute) =>
      _to(route, 'user');

  @override
  void didPop(Route<dynamic> route, Route<dynamic>? previousRoute) =>
      _to(previousRoute, 'user');

  @override
  void didReplace({Route<dynamic>? newRoute, Route<dynamic>? oldRoute}) =>
      _to(newRoute, 'app');

  @override
  void didRemove(Route<dynamic> route, Route<dynamic>? previousRoute) =>
      _to(previousRoute, 'app');
}

// ── Lifecycle ──────────────────────────────────────────────────────────────

/// Records the app going to the background and coming back.
///
/// Its whole job is to give a SILENCE a meaning: thirty seconds spent in
/// another app looks exactly like thirty seconds of reading, and they are
/// opposite kinds of evidence. A distiller working from a voice track over the
/// gap will otherwise narrate the wrong one.
class AlgoLifecycleObserver with WidgetsBindingObserver {
  AlgoLifecycleObserver(this.sinks);
  final Sinks sinks;

  @override
  void didChangeAppLifecycleState(AppLifecycleState state) {
    final hidden = state != AppLifecycleState.resumed;
    sinks.recorder?.visibility(hidden: hidden);
  }
}

/// Register the lifecycle observer and return its teardown.
Unbind bindLifecycle(Sinks sinks) {
  final observer = AlgoLifecycleObserver(sinks);
  final binding = WidgetsBinding.instance..addObserver(observer);
  return () => binding.removeObserver(observer);
}

// ── Network ────────────────────────────────────────────────────────────────

/// Records one FINISHED request. Call this from whatever HTTP client the app
/// uses — a Dio `InterceptorsWrapper`, an `http` `BaseClient`, a hand-rolled
/// wrapper.
///
/// Exposed as a function rather than shipped as a Dio interceptor on purpose:
/// binding to Dio would make it a dependency of this package for every app,
/// including the ones that do not use it. Two lines in the host beats a
/// dependency in everybody's `pubspec.yaml`.
///
/// A success is not evidence — recording every 200 would bury the request that
/// mattered and spend the trace's 500-event budget on noise.
void recordRequest(
  Sinks sinks, {
  required String method,
  required String path,
  required int status,
  num? ms,
}) {
  if (status > 0 && status < 400) return;
  sinks.recorder?.request(method: method, path: path, status: status, ms: ms);
  sinks.crash
      ?.noteFailedRequest(method: method, path: path, status: status, ms: ms);
}

// ── Everything, once ───────────────────────────────────────────────────────

/// Install every binding that does not need the host to hand us something, and
/// return ONE teardown.
///
/// Navigation is absent deliberately: it needs an observer the host adds to its
/// own `MaterialApp`, and an SDK cannot reach in and do that. So is the network
/// interceptor, for the reason above. Both are two lines in the app, and both
/// are named in the install task.
///
/// Teardown runs in reverse install order, so a handler chained on top of
/// another is removed before the one beneath it — otherwise unbinding ours
/// leaves the host's handler pointing at a function we have discarded.
Unbind bindAll({
  required Sinks sinks,
  bool logs = true,
  bool crashes = true,
  bool lifecycle = true,
}) {
  final undo = <Unbind>[];
  if (logs) undo.add(bindDebugPrint(sinks));
  if (lifecycle) undo.add(bindLifecycle(sinks));
  final crash = sinks.crash;
  if (crashes && crash != null) {
    undo.add(bindFlutterErrors(crash));
    undo.add(bindPlatformErrors(crash));
  }
  return () {
    for (final fn in undo.reversed) {
      try {
        fn();
      } catch (_) {
        // A teardown that throws must not strand the others.
      }
    }
    undo.clear();
  };
}
